middleware/proxy.go

```diff
@@ -211,13 +239,29 @@ func (b *roundRobinBalancer) Next(c echo.Context) *ProxyTarget {
 	} else if len(b.targets) == 1 {
 		return b.targets[0]
 	}
-	// reset the index if out of bounds
-	if b.i >= len(b.targets) {
-		b.i = 0
+
+	var i int
+	const lastIdxKey = "_round_robin_last_index"
+	// This request is a retry, start from the index of the previous
+	// target to ensure we don't attempt to retry the request with
+	// the same failed target
+	if c.Get(lastIdxKey) != nil {
```

Comment:
I do not think this is bullet proof. Targets list can be mutated between retries so that index does could point to different target potentially. 

I suggest:

1. add `contextKey string` field to `roundRobinBalancer` struct
2. add method for setting it go-routine safe way
```go

func (b *roundRobinBalancer) SetContextKey(contextKey string) {
	b.mutex.Lock()
	defer b.mutex.Unlock()

	b.contextKey = contextKey
}
```
3. In `ProxyWithConfig` setup part lets set that key when balancer is of type `roundRobinBalancer`
```go
	if b, ok := config.Balancer.(*roundRobinBalancer); ok {
		b.SetContextKey(config.ContextKey)
	}
```
4. and in `roundRobinBalancer.Next` get it from context and choose next target that is not that old one

Reply:
My concern with this is that we only store the last failed target. If multiple targets fail (which I we acknowledge can happen since more than 1 retry is supported) then you may still end up retrying a fail target. Checking for a specific balancer implementation in order to set the key also doesn't feel great.  

You are right though that the first solution is not perfect as the list of targets may change - but for me that feels like the "less bad" solution to me, given that targets shouldn't change that frequently (I assume). 

Another option could be to store a list (or maybe "set" using map[string]bool) of failed targets (or target names?)  Then we can get the best of both worlds. We avoid the potential issue when the target list is changed, and also avoid "forgetting" about previously failed targets (the map would only be initialized on the context when a request failed I assume).

Reply:
I think it has to work correctly - if feature we are implementing here is to ignore last target it must be correct in regards of data-races and using index is not safe/correct.

> implementation in order to set the key also doesn't feel great

this check could be check for anynomous interface
```go
	if b, ok := config.Balancer.(interface{ SetContextKey(string) }); ok {
		b.SetContextKey(config.ContextKey)
	}
```
but I think `.(*roundRobinBalancer)` is OK because problem of passing context key is relevant only for Echo own private balancers which constructor methods we can not change. User defined balancers are created by users and therefore are able to have other means to know context key.

I do not think that choosing data-race over ineloquent cast is correct. If that case why not just assume that context key is not changed - and use always same constant key? These requirements/limitations are us to set at the moment. One day we will get issue/bug related that thing that we just closed our eyes to - it does not make sense to ignore it.

> My concern with this is that we only store the last failed target.

now this is scope creep here. If we start go down this path to ignore all failed targets (not only last one), I think we should not implement that at all in default balancers. This feature cries for request scoped stateful balancer instances.  And eventually leads to question like - what if all fail? What to do with ones what keep failing?  Do we need circuit breaker(s) etc.

Reply:
> I think it has to work correctly - if feature we are implementing here is to ignore last target it must be correct in regards of data-races and using index is not safe/correct.

I wouldn't call it a data-race, personally. It's something that will only happen if the list of targets is changed while a failed request is in the process of being retried. Even if it does happen, the worst-case scenario is for a very short period, a request or two may end up being retried with the same failed target. This could be documented (`if the list of targets is changed while a failed request is being retried, a failed target may be returned by the balancer`), is a reasonable constraint/limitation and would be understandable to users.

_(Incidentially, looking at the existing code after you raised this possibility, there is actually a chance for a slice out of bounds panic in the current round robin implementation if the list of balancers is changed. I'm going to push a commit to fix this as part of this PR. If this has never been reported, it suggests not many users are actually changing the balancer targets at runtime anyway.)_

> now this is scope creep here. If we start go down this path to ignore all failed targets (not only last one), I think we should not implement that at all in default balancers

I disagree that it's scope creep. We already have `RetryCount` defined as a `int`, suggesting to users we will perform multiple retries. It's just making sure the behaviour is correct and meets users expectations. 

From a users point of view, imagine a scenario where there are 3 targets, 2 are failing, 1 is OK,  and `RetryCount` is set to 2 - I would expect all requests will succeed. With the change you are proposing, many requests would not succeed in a concurrent environment as they'd end up retrying a failed target on the second retry. With the change I am proposing, all requests would succeed with the downside that if the target list is changed while failed requests are being retried, there is a possibility for a short time period some of those requests may retry a failed target. 

Apologies, but I'm hesitant to push a change I don't believe is the correct one. I'll push a few more commits to add some documentation and also fix the potential panic with the existing balancer code. If you decide to decline the PR, I totally understand. You could always take my changes as a patch and implement your own solution from there. Appreciate you taking the time to comment and feedback.

Reply:
I do not think that using index where to avoid target is correct. It is a "race" as from the time it is set and by the time it use used - that index can refer to different targets.  I think this PR is otherwise fine but I can not get over this part.

@lammel please be arbitrator here and do a review.